os/os.go

```diff
@@ -0,0 +1,159 @@
+	i.version = version
+	return i
+}
+
+// Standby specifies if supervisor is on standby.
+func (i *InstallOperation) Standby(standby bool) *InstallOperation {
+	i.standby = standby
+	return i
+}
+
+// Reader specifies the package reader for the OS file.
+func (i *InstallOperation) Reader(reader io.Reader) *InstallOperation {
+	i.reader = reader
+	return i
+}
+
+// NewInstallOperation creates an empty InstallOperation.
+func NewInstallOperation() *InstallOperation {
+	return &InstallOperation{}
+}
+
+// awaitPackageInstall receives messages from the client until either
+// (a) the package is installed and validated, in which case it returns the InstallResponse message
+// (b) the device does not have the package, in which case it returns a nil response
+// (c) an error occurs, in which case it returns the error
+func awaitPackageInstall(ic ospb.OS_InstallClient) (*ospb.InstallResponse, error) {
+	for {
+		cresp, err := ic.Recv()
+		if err != nil {
+			return nil, err
+		}
+		switch v := cresp.GetResponse().(type) {
+		case *ospb.InstallResponse_Validated:
+			return cresp, nil
+		case *ospb.InstallResponse_TransferReady:
+			return nil, nil
+		case *ospb.InstallResponse_InstallError:
+			errName := ospb.InstallError_Type_name[int32(v.InstallError.Type)]
+			return nil, fmt.Errorf("installation error %q: %s", errName, v.InstallError.GetDetail())
+		case *ospb.InstallResponse_TransferProgress:
+			log.Infof("installation progress: %v bytes received from client", v.TransferProgress.GetBytesReceived())
+		case *ospb.InstallResponse_SyncProgress:
+			log.Infof("installation progress: %v%% synced from supervisor", v.SyncProgress.GetPercentageTransferred())
+		default:
+			return nil, fmt.Errorf("unexpected client install response: %v (%T)", v, v)
+		}
+	}
+}
+
+func transferContent(ic ospb.OS_InstallClient, reader io.Reader) error {
+	// The gNOI SetPackage operation sets the maximum chunk size at 64K,
+	// so assuming the install operation allows for up to the same size.
+	buf := make([]byte, 64*1024)
+	for {
+		n, err := reader.Read(buf)
+		if n > 0 {
+			if err := ic.Send(&ospb.InstallRequest{
+				Request: &ospb.InstallRequest_TransferContent{
+					TransferContent: buf[0:n],
+				}}); err != nil {
+				return err
+			}
+		}
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return ic.Send(&ospb.InstallRequest{
+		Request: &ospb.InstallRequest_TransferEnd{
+			TransferEnd: &ospb.TransferEnd{},
+		}})
+}
+
+// Execute performs the Install operation.
+func (i *InstallOperation) Execute(ctx context.Context, c *internal.Clients) (*ospb.InstallResponse, error) {
+	ic, icErr := c.OS().Install(ctx)
+	if icErr != nil {
+		return nil, icErr
+	}
+
+	installReq := &ospb.InstallRequest{Request: &ospb.InstallRequest_TransferRequest{TransferRequest: &ospb.TransferRequest{
+		Version:           i.version,
+		StandbySupervisor: i.standby,
+	}}}
+
+	if err := ic.Send(installReq); err != nil {
+		return nil, err
+	}
+
+	select {
```

Comment:
I think we should put the select statement inside a for loop and remove the for loop from awaitPackageInstall. Otherwise, cancelling the context cannot interrupt the install at all. Then awaitPackageInstall should probably be renamed to something without "await" in it. Maybe just "receiveMessage"?

Reply:
It seemed that this would be a issue for transferContent function too.. so I passed the ctx to both functions and check for context cancel there instead of moving the for loop. Please let me know if this looks ok.